### ISSUE TITLE
OM, SM, and GM may have no reference, template, or constant

### ISSUE THREAD

**Issue by chrdebru** [kg-construct/rml-core#111]
https://github.com/chrdebru/rml-core/blob/missing-test/shapes/expression_map.ttl

Implements the following:

```
    # Exactly one rml:template, one rml:constant, one rml:reference,
    # one rml:returnMap, or one rml:functionExecution is required.
```

We, however, allow no such thing for subject maps and object maps (and arguably graph maps) as well. The iterations are then used to generate blank nodes.

**Comment by DylanVanAssche** [comment]
This construct you mention is to enforce that any of these options is picked (so minCount == 1) and also that you don't put multiple of these options (maxCount == 1).

> allow no such thing for subject maps and object maps (and arguably graph maps) 

I cannot follow here. Each of these maps must have exactly one of these options right? A subject map must have a template/constant/reference/... right?
You can enforce it more strict in the specific maps though if needed.

**Comment by DylanVanAssche** [comment]
Rethinking this issue: is this when you have `rml:termType rml:BlankNode` without a template?
If that's the only 'exception' to the rule, we should maybe define this restriction:

```
IF termType == BlankNode
  minCount = 0
ELSE
  minCount = 1
```

?

**Comment by chrdebru** [comment]
Yes, but what if no `rml:termType` is provided? I'm also thinking about the Gather Maps, and it is difficult (impossible?) to "overwrite" shapes in SHACL. My approach is more: if no `rml:template`, `rml:reference`, and `rml:constant`, then the `rml:termType` is assumed to be `rml:BlankNode`. This rule would work with predicate maps as well as the inferred term type blank node violates the `rml:termType rr:IRI` constraint.

**Comment by DylanVanAssche** [comment]
> Yes, but what if no rml:termType is provided? 

There should be a default which should be used. For SubjectMaps this is IRI. It will be a 'fun' job to actually check this in SHACL shapes...

> and it is difficult (impossible?) to "overwrite" shapes in SHACL.

I'm not sure how you can overwrite things I'm afraid... It is purely an AND/OR/XOR story AFAIK. Correct me if I'm wrong.

**Comment by chrdebru** [comment]
I would argue that the default when no constant, template, or reference is provided should be blank nodes. That's the only thing that you can generate based on the iteration. Defaults can be "inferred" once you have one of them.

**Comment by dachafra** [comment]
This depends on the final decision of #52

**Comment by dachafra** [comment]
> I might have missed something, but, for me, a TM must have exactly one SM and zero or more POMs. 
True 😅

> When the SM does not have a constant, reference, or template, then a BN identifier based on the iteration is created. The same thing for OM. I don't see how removing the min count leads to TMs without SMs.

So a potential example of TM with this option could be (empty `SM`):

```turtle
@prefix rml: <http://w3id.org/rml/> .
@prefix ex: <http://example.com/ns#> .

<#ImageLogicalSource>
  rml:source [
    ...
  ] ;
  rml:iterator "$.Images[*]" ;
  rml:referenceFormulation rml:JSONPath .

<#ImageTriplesMap>
  rml:logicalSource <#ImageLogicalSource> ;
  rml:subjectMap [] ;
  rml:predicateObjectMap [
    rml:predicate ex:height ;
    rml:objectMap [
      rml:reference "$.Height" ;
    ] ;
  ] .
```

**Comment by pmaria** [comment]
Why would we not allow omitting the subject map in this case?

**Comment by chrdebru** [comment]
So the default SM is `[]` (no expression map, no class declarations, no graph maps, no term type, ...) unless explicitly declared? That means that we have to specify that for each "non-occurrence", there is a new blank SM. That works, yes. My suggestion keeps the original cardinality constraint.
